working/333 - shared memory multithreading/proposal.md

```diff
@@ -1501,29 +1161,29 @@ access the value.
 > `std::atomic<int32_t>` (C++) or `_Atomic int32_t` (C11) because these types
 > don't have a defined ABI.
 
-## Prototyping Roadmap
-
-The change of this impact has to be carefully evaluated. I suggest we start with
-bare minimum needed to validate share memory concurrency in a realistic setting:
+## Implementation Roadmap
 
-1. Implement `shared` global fields using a VM specific pragma
-   `@pragma('vm:shared')`.
-2. Hiding under an experimental flag and in a separate library (e.g.
-   `dart:concurrent`):
+We start by implementing _shared isolates_ and allowing `shared` global
+fields (designated via `@pragma('vm:shared')` rather than a keyword) of
+trivially shareable types. We then expose shared isolates to FFI by introducing
+`NativeCallable.shared` and allowing to call into an isolate group from
+an arbitrary thread.
 
-- Introduce `Shareable` interface and enforce suggested restrictions using a
-  Kernel transformation instead of incorporating them into CFE.
-- Introduce minimum amount of core library changes (e.g. `ShareableList`).
+These changes do not significantly change the shape of Dart programming
+language, they streamline the interoperability with native code but do not
+introduce any new fundamental capabilities: developers can already share
+native memory between isolates and that simply makes such sharing more
+convenient to use. There is no sharing of mutable Dart objects at this stage
+yet.
```

Comment:
Actually a very common use case in C interop will be to pass a closure to `NativeCallable.shared()` and make that closure close over arbitrary state. We'd want to use the same closure (and not necessarily deeply clone it) thereby possibly sharing mutable objects.

(If we do copy the closure things get complicated, because the callback may be executed on N threads in parallel then we'd need N copies of the closure - i.e. it's not sufficient to deeply clone the closure once at `NativeCallable.shared()` call time but we'd need to do that every time it's invoked)

Would it make sense to have this API under the experimental flag?

Reply:
I have added the following restriction to `NativeCallable.shared`:

> ```dart
>  /// Throws [ArgumentError] if [callback] captures state which can't be
>  /// transferred to shared isolate without copying.
>  external factory NativeCallable.shared(...);
> ```
> ...
> In _shared **everything** multithreading_ world `callback` can be allowed to
> capture arbitrary state, however in _shared **native memory** multithreading_
> this state has to be restricted to trivially shareable types

I would like to avoid creating mutable shared _Dart_ state between initially.